doc/source/dev/api-dev/array_api.rst

```diff
@@ -0,0 +1,253 @@
+PyTorch tensors as inputs and return values:
+
+.. code:: python
+
+    >>> import torch
+    >>> from scipy.cluster.vq import vq
+    >>> code_book = torch.tensor([[1., 1., 1.],
+    ...                           [2., 2., 2.]])
+    >>> features  = torch.tensor([[1.9, 2.3, 1.7],
+    ...                           [1.5, 2.5, 2.2],
+    ...                           [0.8, 0.6, 1.7]])
+    >>> code, dist = vq(features, code_book)
+    >>> code
+    tensor([1, 1, 0], dtype=torch.int32)
+    >>> dist
+    tensor([0.4359, 0.7348, 0.8307])
+
+Note that the above example works for PyTorch CPU tensors. For GPU tensors or
+CuPy arrays, the expected result for ``vq`` is a ``TypeError``, because ``vq``
+is not a pure Python function and hence won't work on GPU.
+
+More strict array input validation will reject ``np.matrix`` and
+``np.ma.MaskedArray`` instances, as well as arrays with ``object`` dtype:
+
+.. code:: python
+
+    >>> import numpy as np
+    >>> from scipy.cluster.vq import vq
+    >>> code_book = np.array([[1., 1., 1.],
+    ...                       [2., 2., 2.]])
+    >>> features  = np.array([[1.9, 2.3, 1.7],
+    ...                       [1.5, 2.5, 2.2],
+    ...                       [0.8, 0.6, 1.7]])
+    >>> vq(features, code_book)
+    (array([1, 1, 0], dtype=int32), array([0.43588989, 0.73484692, 0.83066239]))
+
+    >>> # The above uses numpy arrays; trying to use np.matrix instances or object
+    >>> # arrays instead will yield an exception with `SCIPY_ARRAY_API=1`:
+    >>> vq(np.asmatrix(features), code_book)
+    ...
+    TypeError: 'numpy.matrix' are not supported
+
+    >>> vq(np.ma.asarray(features), code_book)
+    ...
+    TypeError: 'numpy.ma.MaskedArray' are not supported
+
+    >>> vq(features.astype(np.object_), code_book)
+    ...
+    TypeError: object arrays are not supported
+
+
+Currently supported functionality
+`````````````````````````````````
+
+The following modules provide Array API standard support when the environment
+variable is set:
+
+- ``scipy.cluster.hierarchy``
+- ``scipy.cluster.vq``
+
+
+Implementation notes
+--------------------
+
+A key part of the support for the array API standard and specific compatibility
+functions for Numpy, CuPy and PyTorch is provided through
+`array-api-compat <https://github.com/data-apis/array-api-compat>`_.
+This package is included in the SciPy code base via a git submodule (under
+``scipy/_lib``), so no new dependencies are introduced.
+
+``array-api_compat`` provides generic utility functions and adds aliases such
+as ``xp.concat`` (which, for numpy, maps to ``np.concatenate``). This allows
+using a uniform API across NumPy, PyTorch and CuPy (as of right now; support
+for other libraries like JAX is expected to be added in the future).
+
+When the environment variable isn't set and hence Array API support in SciPy is
+disabled, we still use the "augmented" version of the NumPy namespace, which is
```

Comment:
What does that mean? Will it accept a pytorch tensor? Will it convert the tensor to a numpy array?

Reply:
If the flag is off, it's as is we were not doing anything (so no, it only accepts NumPy arrays). Except that we don't use the plain NumPy namespace but we use `array_api_compat.numpy`. This is because as we update the code to the standard, we need functions from the standard like `xp.concat` which is not present in plain NumPy.